### PR TITLE
76440 - 76443 - create upload aio zip to fss with enc exchange set and large media exchange set

### PR DESCRIPTION
**Pull request by shailen14599**
This PR includes PBI#76440 and PBI#76443 changes

Added code changes for AIO.zip creation with ENC exchange set and large media exchange set
Added Unit Test cases
Added devops changes for FT's in QC stage
Added Functional tests

### CODE REVIEW

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs [hunk 1]
```diff
@@ -25,9 +25,11 @@ public class FulfilmentFileShareServiceTest
         public bool fakeIsZipFileCreated = false;
         private bool fakeIsBatchCommitted = false;
         public string fakeExchangeSetRootPath = @"D:\\Downloads\";
-        public string fakeBatchId = "7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272";
+        readonly string fakeBatchId = "7b4cdf10-adfa-4ed6-b2fe-d1543d8b7272";
```
Comment (DanOwenUKHO) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:28]: Add access modifier

**Review by DanOwenUKHO — CHANGES_REQUESTED** [review]
A few minor changes.

The final mutation score is 51.86 %

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/FulfilmentServiceJob.cs [hunk 2]
```diff
@@ -111,6 +111,7 @@ public async Task ProcessQueueMessage([QueueTrigger("%ESSFulfilmentStorageConfig
 
         public async Task CreateAndUploadErrorFileToFileShareService(SalesCatalogueServiceResponseQueueMessage fulfilmentServiceQueueMessage, EventId eventId, string errorMessage, string batchFolderPath)
         {
+            var isErrorFileCommitted = false;
```
[the remark this answers is not in the document]
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.FulfilmentService/FulfilmentServiceJob.cs]: resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 1 above)
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:28]: strange but access modifier is already present in this line.

Comment (DanOwenUKHO) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs:28]: For me the line is missing a private/internal/protected/public modifier

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled.cs [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+using System.Threading.Tasks;
+using UKHO.ExchangeSetService.API.FunctionalTests.Helper;
+using UKHO.ExchangeSetService.API.FunctionalTests.Models;
+
+namespace UKHO.ExchangeSetService.API.FunctionalTests.FunctionalTests
+{
+    [TestFixture]
+    class ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled
+    {
+        private string EssJwtToken { get; set; }
```
[the remark this answers is not in the document]
Comment (Arit14975) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled.cs:15]: Hi Dan, Your comment is wise as we are working on existing project's there are already 40+ existing FT's. So for the new FT's/ .cs file as well we are following the same approach. Thanks

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs (hunk 1 above)
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentFileShareServiceTest.cs]: Resolved

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ray-sj) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled.cs:15]: so there is nothing wrong with making improvements so Dan is write we and start to implement a common project and retro fit when needed the common points

**Review by ray-sj — CHANGES_REQUESTED** [review]
in general looks fine, but
1. static waits need to be reviewed as these slow the test execution down and are a constant regardless of the system response
2. common functionality and variables need to be grouped and not duplicated as this leeds to missing them in the future and multiple points of failure if updates etc.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/AzureADB2CAuthenticationTests.cs [hunk 4]
```diff
@@ -23,6 +23,7 @@ class AzureADB2CAuthenticationTests
         [SetUp]
         public async Task SetupAsync()
         {
+            await Task.Delay(30000);
```
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/AzureADB2CAuthenticationTests.cs]: Hi Ray,

As discussed and agrees, we have updated the test cases as For this project, we have to introduced new FT categories. 
**SmokeTest-AioDisabled, QCOnlyTest-AioDisabled**
These categories are used for the existing FTs which will be treated as Regression Test cases when AIO is Disabled

**SmokeTest-AioEnabled, QCOnlyTest-AioEnabled**
These categories are used for the new FTs which when AIO is Enabled

In order to run the FTs in pipeline based on the above categories, we have introduced swapping of the API config in the pipeline and re-run the API. As the system needs some seconds to start working again, we don't have anything were we can check the status of it so can't have a dynamic wait. This is the reason we have introduced the static wait of 30sec, which we observed a minimal wait in order to run the FTs successfully. 

Before coming to this solution we connected with Dev & Dev_ops too and checked if they can handle this 30 sec in any of the deployment step but the answer was No. This is the another reason for adding it in the Fts.

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mayuresh-Gawde) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.API.FunctionalTests/FunctionalTests/ExchangeSetGenerateFilesForAioProductIdentifierWhenAioIsEnabled.cs:15]: Hi Ray,

As discussed and agreed, we will take this in the PBI 74332: ESS: Automated Functional Tests when AIO feature is toggle ON

**Review by ray-sj — APPROVED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentDataServiceTest.cs [hunk 5]
```diff
@@ -293,6 +296,33 @@ private SalesCatalogueDataResponse GetSalesCatalogueDataResponseForAio()
...
+        private IDirectoryInfo GetSubDirectories(string exchangeSetType)
+        {
+            IDirectoryInfo fakeDirectoryInfo = A.Fake<IDirectoryInfo>();
+            if (exchangeSetType == EncExchangeSet)
+            {
+                A.CallTo(() => fakeDirectoryInfo.Name).Returns(fakeFileShareServiceConfig.Value.ExchangeSetFileFolder);
+                A.CallTo(() => fakeDirectoryInfo.FullName).Returns(fakeFileShareServiceConfig.Value.ExchangeSetFileFolder);
+                A.CallTo(() => fakeDirectoryInfo.GetDirectories()).Returns(new IDirectoryInfo[] { fakeDirectoryInfo });
+            }
+            else if (exchangeSetType == "AIO")
```
Comment (harshal11869) [inline UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentDataServiceTest.cs:308]:
```suggestion
            else if (exchangeSetType == AioExchangeSet)
```

**Review by harshal11869 — CHANGES_REQUESTED** [review]

File: UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentDataServiceTest.cs (hunk 5 above)
Comment (shailen14599) [reply UKHO.ExchangeSetService.API/UKHO.ExchangeSetService.Webjob.UnitTests/Services/FulfilmentDataServiceTest.cs]: Resolved

**Review by DanOwenUKHO — APPROVED** [review]

**Review by amogh14893 — APPROVED** [review]
Verified DevOps Changes ,Looks good to me

**Review by harshal11869 — APPROVED** [review]
Looks Good.

**Pull request merged**